Add test for JHOVE module validation

Dhwaniartefact opened the pull request:
This PR adds acceptance test for JHOVE modules validation.

Connected to: https://github.com/archivematica/Issues/issues/1723

Dhwaniartefact committed:
Add test for Jhove module validation

replaceafill commented on features/black_box/validation.feature:
```diff
@@ -0,0 +1,11 @@
+    And the "Validate formats" job fails
+    And 16 "Validate formats" tasks were executed
+    And 8 "Validate formats" tasks failed
+    And 8 "Validate formats" were successful
```
I think this is missing `tasks`. What do you think about?

```suggestion
    And 8 "Validate formats" tasks succeeded
```

The step helper needs to be updated accordingly.

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1528,42 @@ def step(context):
+
+
+@then('{task_count} "{job_name}" tasks were executed')
+def step_impl(context, task_count: Optional[int] = None, job_name: str = "") -> None:
```
Why is the `task_count` optional? Could you provide an example of how the step would like in the feature file in that case?

Also, why is this the only new step with type annotations?

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1528,42 @@ def step(context):
+
+@then('{task_count} "{job_name}" tasks were executed')
+def step_impl(context, task_count: Optional[int] = None, job_name: str = "") -> None:
+    unit_uuid = context.current_transfer["transfer_uuid"]
+    jobs = utils.get_jobs(context.api_clients_config, unit_uuid, job_name=job_name)
```
Would this prevent the step from being used in the `Ingest` phase where `sip_uuid` is used?

Same observation for the other steps.

Dhwaniartefact replied:
yes for `ingest` phase, `sip_uuid` is going to be used.

Dhwaniartefact replied:
Currently this test has limitations of using it into `transfer` phase. Let me make this more generic. Thanks for the suggestion!

replaceafill replied:
Since this limitation currently doesn't block you (as far as I can tell) from developing the other steps we've agreed on implementing, I'd recommend leaving this `generalization` part for last if that works better for you.

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1528,42 @@ def step(context):
+    assert len(jobs), f"No jobs found for unit {unit_uuid}"
+
+    fail_task = 0
+    for job in jobs:
+        for task in job["tasks"]:
+            if task["exit_code"] == 1:
+                fail_task += 1
+        assert fail_task == int(task_count)
```
Suppose I have three "some job"s:

- First "some job": 3 tasks with exit codes: 0, 0, 1
- Second "some job": 3 tasks with exit codes: 1, 0, 0
- Third "some job": 2 tasks with exit codes: 1, 1

And I write the step as:

```gherkin
   And 4 "some job" tasks failed`
```

Would this step pass with that scenario?

Also in the `int(task_count)` you're expecting an `int`. Behave provides converters for doing that. See the `event_count` parameter in: https://github.com/artefactual-labs/archivematica-acceptance-tests/blob/255cca0f129230cbf9102305895d82f0ca8f6f16/features/steps/black_box_steps.py#L409-L413 or the `expected_files_count` parameter in: https://github.com/artefactual-labs/archivematica-acceptance-tests/blob/255cca0f129230cbf9102305895d82f0ca8f6f16/features/steps/black_box_steps.py#L416-L442

This functionality is based on https://pypi.org/project/parse/

replaceafill reviewed (commented):
@Dhwaniartefact Here's some initial feedback while I work on fixing the test workflow.

Dhwaniartefact committed:
Add suggested changes

Dhwaniartefact committed:
Add test for JHOVE validation

replaceafill commented on features/black_box/validation.feature:
```diff
@@ -0,0 +1,45 @@
+    And 1 WARC file is failed
+    And 1 WARC file is succeeded
+    And 1 WAVE file is failed
+    And 1 WAVE file is succeeded
```
These steps are tightly coupled to the `JHOVEModulesValidation` sample transfer so I think they can be moved to the `Scenario` above. At that point this `Scenario Outline` loses value because we're not asserting anything about JHOVE modules for other transfers. So I'd propose to move these (and other relevant) steps and remove the `Scenario Outline`.

Dhwaniartefact replied:
Thanks for noting this. I have added this change into my next [commit](https://github.com/artefactual-labs/archivematica-acceptance-tests/blob/5ebe66e174ae1808f62da68da9362828ca22b3f9/features/black_box/validation.feature).

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1527,78 @@ def step(context):
         assert uses_order_indexes == sorted(uses_order_indexes), error
+
+
+def get_uuid_value(context, unit_type):
```
```suggestion
def get_unit_uuid(context, unit_type):
```

Dhwaniartefact replied:
Thanks for suggestion. I changed into my next commit.

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1527,78 @@ def step(context):
+    unit_uuid = get_uuid_value(context, unit_type)
+    jobs = utils.get_jobs(context.api_clients_config, unit_uuid, job_name=job_name)
+    assert len(jobs), f"No jobs found for unit {unit_uuid}"
+    assert len(jobs[0]["tasks"]) == task_count
```
Why does this step only consider the tasks of the first job (`jobs[0]["tasks"]`) while the other ones iterate through all of them?

Dhwaniartefact replied:
I have added iteration into my [next commit.](https://github.com/artefactual-labs/archivematica-acceptance-tests/blob/11efff96a77861a0a6034613acb742c4773df39b/features/steps/black_box_steps.py#L1546-L1552)

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1527,78 @@ def step(context):
+
+@then("{file_count:d} {file_extension} file is {status}")
+def step_impl(context, file_count, file_extension, status):
+    unit_uuid = context.current_transfer["transfer_uuid"] + "/?detailed=1"
```
Could we start using https://github.com/artefactual-labs/amclient/commit/b34ae76a584b5014d24adbc21067e3d0218a2a93 to do this? You'd need to modify the `get_jobs` helper to get a parameter to request detailed task output from `AMClient`.

Dhwaniartefact replied:
Yes, I have added this commit [artefactual-labs/amclient@b34ae76](https://github.com/artefactual-labs/amclient/commit/b34ae76a584b5014d24adbc21067e3d0218a2a93)  into `requirements.in` file and modify `get_jobs` helper to fetch detailed task output into upcoming commit.

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1527,78 @@ def step(context):
+
+
+@then('{task_count:d} "{job_name}" {unit_type} tasks failed')
+def step_impl(context, job_name, task_count, unit_type):
```
```suggestion
def step_impl(context, task_count, job_name, unit_type):
```

Dhwaniartefact replied:
Thanks for suggestion. I changed into my next commit.

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1527,78 @@ def step(context):
+
+
+@then('{task_count:d} "{job_name}" {unit_type} tasks succeeded')
+def step_impl(context, job_name, task_count, unit_type):
```
```suggestion
def step_impl(context, task_count, job_name, unit_type):
```

Dhwaniartefact replied:
Thanks for suggestion. I changed into my next commit.

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1527,78 @@ def step(context):
+
+    for job in jobs:
+        for task in job["tasks"]:
+            if status == "failed":
```
Suppose there's a typo in the status in one of the steps, so the feature file was written like this:

```gherkin
    And 1 AIFF file is faiLed
```

Neither `if` is going to detect that and the step won't assert anything which might seem like correctness.

I think before the loop you need to assert the `status` value is something you can handle or return an error accordingly.

Dhwaniartefact replied:
sure let me assert `status` value well in advance before we start iteration for tasks.

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1527,78 @@ def step(context):
+    for job in jobs:
+        for task in job["tasks"]:
+            if status == "failed":
+                if (file_extension).lower() == task["file_name"].split(".")[
+                    -1
+                ] and task["exit_code"] == 1:
```
Suppose that the transfer that uses this step with a `PDF` extension contains two files: `myfile.pdf` and `MYOTHERFILE.PDF`.

If you use `lower()` on the `file_extension`, should you also lower the `task["file_name"]`?

And for getting the extension of the `task["file_name"]` I'd recommend using Python's [os.path.splitext](https://docs.python.org/3/library/os.path.html#os.path.splitext) or [Path.suffix](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.suffix).

Dhwaniartefact replied:
> And for getting the extension of the `task["file_name"]` I'd recommend using Python's [os.path.splitext](https://docs.python.org/3/library/os.path.html#os.path.splitext) or [Path.suffix](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.suffix).

In my next commit, I have used `pathlib` module to get the value of `file_extension`. That's really helped. Thanks @replaceafill

Dhwaniartefact replied:
> Suppose that the transfer that uses this step with a `PDF` extension contains two files: `myfile.pdf` and `MYOTHERFILE.PDF`.
> 
> If you use `lower()` on the `file_extension`, should you also lower the `task["file_name"]`?

Got your point and made changes accordingly into next commit.

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1527,78 @@ def step(context):
+                    -1
+                ] and task["exit_code"] == 1:
+                    fail_file += 1
+                    assert file_count == fail_file
```
Would this assert work if I have three failing JPEG files and I write the step like:

```gherkin
And 3 JPG file is failed
```

I think in the first iteration this would initially compare `assert 3 == 1` which will immediately stop? I think it has been working so far because you feature file only has 1s in all your steps.

Also it'd be nice if all the `assert ...` statements you add have descriptive error messages that give the developer details of what went wrong.

Dhwaniartefact replied:
> 
> Also it'd be nice if all the `assert ...` statements you add have descriptive error messages that give the developer details of what went wrong.

Yes, I have added error messages for each assert statement to provide better clarity and understanding for developers.

replaceafill replied:
I commented on all the new error messages. I think they're inaccurate in some situations.

Dhwaniartefact replied:
oops, let me go through them

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1527,78 @@ def step(context):
+                    assert file_count == fail_file
+                else:
+                    continue
+            if status == "succeeded":
```
I don't think the step needs to have two separate `if`s for the `status` you want to handle.

See https://github.com/artefactual-labs/archivematica-acceptance-tests/blob/c9b1705cef238d84f8da8cbaa658fe613334b0e4/features/steps/black_box_steps.py#L535-L553

I think you could create a mapping for the `status` values you expect and map them to the `exit_code` of the tasks. Something like:

```python
status_to_task_exit_code = {
    "failed": 1,
    "succeeded": 0,
}
```

Then you could use that mapping to assert the `status` that was passed to you:

```python
assert status in status_to_task_exit_code, "descriptive error message"
```

as I recommended in a previous comment (before the loop).

That mapping would also allow you to get the `exit_code` you expect from the `status`:

```python
expected_exit_code =  status_to_task_exit_code[status]
```

and use it in the last bit of your `if` instead of hard coding the values:

```python
and task["exit_code"] == expected_exit_code
```

Tasks could also have other exit codes, like (`179` to indicate failure](https://github.com/artefactual/archivematica/blob/61e5ca1422b2fe79e626975c6483af0f25486416/src/MCPServer/lib/assets/workflow.json#L4596-L4599).

A better approach could be a mapping of tuples like the example above:

```python
status_to_task_exit_codes = {
    "failed": (1, 179),
    "succeeded": (0,),
}
```

changing also the other bits to:

```python
expected_exit_codes =  status_to_task_exit_codes[status]
...
and task["exit_code"] in expected_exit_codes
```

I think this approach also makes the two separate variables `success_file` and `fail_file` irrelevant and would allow you to have a single `assert ...` on a total of tasks as the last line of the step after the loop is done.

Dhwaniartefact replied:
Thank you for helping me understand this in such a detailed and clear way. 

I try my best to cover this as much as I can in my next [commit.](https://github.com/artefactual-labs/archivematica-acceptance-tests/blob/11efff96a77861a0a6034613acb742c4773df39b/features/steps/black_box_steps.py#L1582-L1618)

replaceafill reviewed (changes requested):
Great progress @Dhwaniartefact. I think I've spotted some potential problems and suggested some ways to address them.

Dhwaniartefact committed:
Update requirements

This adds a amclient commit which introduces `detailed` parameter
in the Jobs API endpoint

Dhwaniartefact committed:
Remove scenario outline from test

Dhwaniartefact committed:
Assert task count and add error messages

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1528,91 @@ def step(context):
+        for task in job["tasks"]:
+            if task:
+                task_size += 1
+    assert task_size == task_count, f"No tasks found for unit {unit_uuid}"
```
Suppose the expected `task_count` is `10` but the iteration calculated a `task_size` of `5`. Is `No tasks found for unit ...` an accurate error message in that case?

Dhwaniartefact replied:
Thanks for suggestion, let me change this.

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1528,91 @@ def step(context):
+        for task in job["tasks"]:
+            if task["exit_code"] == 1:
+                fail_task += 1
+    assert fail_task == task_count, f"No failed task found for unit {unit_uuid}"
```
Suppose the expected `task_count` is `10` but the iteration calculated a `fail_task` of `5`. Is `No failed task found for unit ...` an accurate error message in that case?

Dhwaniartefact replied:
Thanks for suggestion, let me change this.

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1528,91 @@ def step(context):
+        for task in job["tasks"]:
+            if task["exit_code"] == 0:
+                success_task += 1
+    assert success_task == task_count, f"No successful task found for unit {unit_uuid}"
```
Suppose the expected `task_count` is `10` but the iteration calculated a `success_task` of `5`. Is `No successful task found for unit ...` an accurate error message in that case?

Dhwaniartefact replied:
Thanks for suggestion, let me change this.

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1528,91 @@ def step(context):
+    task_size = 0
+    for job in jobs:
+        for task in job["tasks"]:
+            if task:
```
Why is this condition needed?

Dhwaniartefact replied:
Thanks for highlighting this. Let me address this in upcoming commit.

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1528,91 @@ def step(context):
+        "failed": (1, 179),
+        "succeeded": (0,),
+    }
+    assert status in status_to_task_exit_codes, (
+        f"No {status} found for {file_extension} file"
+    )
```
Suppose `faiLed` is the requested `status` and `PDF` the requested `file_extension`. Is `No faiLed found for PDF file` an accurate error message in that case?

Dhwaniartefact replied:
Thanks for suggestion, let me change this.

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1528,91 @@ def step(context):
+    assert success_task == task_count, f"No successful task found for unit {unit_uuid}"
+
+
+def verify_task_exit_code(file_extension, task, expected_exit_codes):
```
Since the function doesn't have a docstring to explain what it's doing I need to use its name to try to understand it. To me it indicates that is going to verify the task exit code. The `expected_exit_codes` parameter makes sense for that and `task` as well. But why is the `file_extension` necessary to verify the "task exit code"?

Dhwaniartefact replied:
I just removed this function since it is calling only once. Not needed anymore.

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1528,91 @@ def step(context):
+        "." + file_extension.lower() == (pathlib.Path(task["file_name"]).suffix).lower()
+        and task["exit_code"] in expected_exit_codes
+    ):
+        return True
```
This pattern looks a bit strange. It checks for a couple of conditions in the `if`. If those conditions are "true-ish" it returns `True`, otherwise `None`.

What do you think of returning the actual condition evaluation?

So instead of doing:

```python
def is_green_paint(...):
    if is_paint(...) and is_green(...):
        return True
```

We do:

```python
def is_green_paint(...):
    return is_paint(...) and is_green(...)
```

Dhwaniartefact replied:
got it. Let me make changes accordingly in next commit.

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1528,91 @@ def step(context):
+        for task in job["tasks"]:
+            if verify_task_exit_code(file_extension, task, expected_exit_codes):
+                total += 1
+        assert file_count == total, (
+            f"No {file_extension} file is {status} for unit {unit_uuid}"
+        )
```
Suppose the expected `status` is `failed`, the expected `file_extension` is `PDF` and the expected `file_count` is `10` but the iteration calculated a `total` of `5`. Is `No PDF file is failed for unit ...` an accurate error message in that case?

Also, suppose I have three `Validation` jobs with tasks:

- First job:
  * task with exit code 1 for a PDF file
  * task with exit code 0 for a PDF file
  * task with exit code 0 for a PDF file
- Second job:
  * task with exit code 0 for a PDF file
  * task with exit code 1 for a PDF file
- Third job:
  * task with exit code 1 for a PDF file
  * task with exit code 1 for a PDF file
  * task with exit code 1 for a PDF file
  * task with exit code 0 for a PDF file

And I write the step as:

```gherkin
   And 5 PDF file is failed
```

Would this step pass with that scenario?

Dhwaniartefact replied:
> Suppose the expected `status` is `failed`, the expected `file_extension` is `PDF` and the expected `file_count` is `10` but the iteration calculated a `total` of `5`. Is `No PDF file is failed for unit ...` an accurate error message in that case?
> 

Not really, got it. Let me rephrase this message.

replaceafill commented on features/steps/utils.py:
```diff
@@ -880,6 +881,8 @@ def get_jobs(
         am.job_link_uuid = job_link_uuid
     if job_name is not None:
         am.job_name = job_name
+    if detailed_task is not None:
+        am.detailed = detailed_task
```
:tada:

replaceafill commented on features/black_box/validation.feature:
```diff
@@ -0,0 +1,32 @@
+    And 8 "Validate formats" transfer tasks failed
+    And 9 "Validate formats" transfer tasks succeeded
+    And the "Validation" microservice is executed
+    And 1 AIFF file is failed
+    And 1 AIFF file is succeeded
+    And 1 GIF file is failed
+    And 1 GIF file is succeeded
+    And 1 JP2 file is failed
+    And 1 JP2 file is succeeded
+    And 1 JPG file is failed
+    And 1 JPG file is succeeded
+    And 1 PDF file is failed
+    And 1 PDF file is succeeded
+    And 1 TIF file is failed
+    And 1 TIF file is succeeded
+    And 1 WARC file is failed
+    And 1 WARC file is succeeded
+    And 1 WAV file is failed
+    And 1 WAV file is succeeded
```
@sromkey Could you please help us with some wording for this step?

We're trying to represent the amount of tasks that either passed or failed for a particular file extension in the `Validate formats` job.

replaceafill replied:
In https://github.com/artefactual-labs/archivematica-acceptance-tests/pull/216#discussion_r1968387243 I wrote an example of a step with multiple files expected which currently reads:

```
   And 5 PDF file is failed
```

sromkey replied:
From our discussion- try removing "is" but pluralize "files" when necessary, e.g.

`And 5 PDF files failed`

`And 1 JP2 file succeeded`

Dhwaniartefact replied:
We agreed on using `And 5 PDF file(s) failed`.

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1528,91 @@ def step(context):
+    jobs = utils.get_jobs(
+        context.api_clients_config,
+        unit_uuid,
+        job_microservice="Validation",
```
After some testing I realized using only the `Validation` microservice is a bit risky because of the `Perform policy checks on originals?` processing configuration setting.

See how that microservice looks when you turn it on:

![Captura desde 2025-02-24 14-50-05](https://github.com/user-attachments/assets/393d2c0b-dda6-47c8-8f1b-3859460932b9)

I think in that case `jobs` would include tasks from the `Policy checks for originals` job which you probably don't want?

Dhwaniartefact replied:
Yes, I just need task details of `Validate formats` jobs. 
So in place of `job_microservice="Validation"`, I changed this to `job_name="Validate formats"` parameter in `get_jobs()` API.

Dhwaniartefact replied:
> So in place of `job_microservice="Validation"`, I changed this to `job_name="Validate formats"` parameter in `get_jobs()` API.

It seems to add `job_name` and `job_microservice` details is a good approach.

replaceafill commented on requirements-dev.txt:
```diff
@@ -4,29 +4,29 @@
-amclient==1.4.0
-    # via -r requirements.txt
+amclient @ git+https://github.com/artefactual-labs/amclient.git@b34ae76a584b5014d24adbc21067e3d0218a2a93
+    # via -r /src/requirements.txt
```
I wonder why we hadn't seen these `/src/` changes before. It might have to do with the target in `hack/Makefile`. Nothing to worry for now though.

Dhwaniartefact committed:
Add suggested error messages and change parameter in get jobs

Dhwaniartefact committed:
Add some more changes

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1545,9 +1545,8 @@ def step_impl(context, task_count, job_name, unit_type):
-        for task in job["tasks"]:
-            if task:
-                task_size += 1
+        for _task in job["tasks"]:
```
Not sure if you're prepending the variable name with the `_...` to avoid a linter to complain about an unused variable. If that's the case it's great. But it might also indicate that you don't need the inner `for` loop at all?

Would this be more appropriate?

```python
task_size = 0
for job in jobs:
    task_size += len(job["tasks"])
```

Or going a bit further:

```python
task_size = sum([len(job["tasks"]) for job in jobs])
```

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1528,93 @@ def step(context):
+    for job in jobs:
+        for task in job["tasks"]:
+            if (
+                "." + file_extension.lower()
```
```suggestion
                f".{file_extension.lower()}"
```

replaceafill commented on features/steps/black_box_steps.py:
```diff
@@ -1527,3 +1528,93 @@ def step(context):
+            ):
+                total += 1
+    assert file_count == total, (
+        f"The expected file count for unit {unit_uuid} is {file_count}, but found {total}. "
```
```suggestion
        f"Expected {file_count} {file_extension} file(s) {status} during file format validation for unit {unit_uuid}, but got {total} instead."
```

replaceafill reviewed (approved):
Left a couple of suggestions, but after that this should be ready. Thank you @Dhwaniartefact.

Dhwaniartefact committed:
Fix the suggested changes
